integration-tests/tests/src/tests/results.ts

```diff
@@ -0,0 +1,1067 @@
+
+      this.realm.objects("TestObject").addListener((testObjects, changes) => {
+        if (firstCall) {
+          expect(testObjects.length).equals(3);
+          expect(changes.insertions.length).equals(0);
+          //@ts-expect-error Deprecated field
+          expect(changes.modifications.length).equals(0);
+          expect(changes.newModifications.length).equals(0);
+          expect(changes.oldModifications.length).equals(0);
+          firstCall = false;
+        } else {
+          expect(testObjects.length).equals(4);
+          expect(changes.insertions.length).equals(1);
+          //@ts-expect-error Deprecated field
+          expect(changes.modifications.length).equals(1);
+          expect(changes.newModifications.length).equals(1);
+          expect(changes.oldModifications.length).equals(1);
+        }
+
+        if (resolve) {
+          resolve();
+        }
+      });
+
+      return new Promise((resolveFunction) => {
+        resolve = resolveFunction;
+        this.realm.write(() => {
+          this.realm.objects<TestObject>("TestObject")[0].doubleCol = 5;
+          this.realm.create("TestObject", { doubleCol: 1 });
+        });
+      }).finally(() => this.realm.objects("TestObject").removeAllListeners());
+    });
+
+    it("should be read-only", function (this: RealmContext) {
+      const objects = this.realm.objects<TestObject>("TestObject");
+
+      this.realm.write(() => {
+        this.realm.create<TestObject>("TestObject", { doubleCol: 1 });
+      });
+
+      expect(() => {
+        //@ts-expect-error Should be an invalid write to read-only object.
+        objects[-1] = { doubleCol: 0 };
+      }).throws("Index -1 cannot be less than zero.");
+      expect(() => {
+        //@ts-expect-error Should be an invalid write to read-only object.
+        objects[0] = { doubleCol: 0 };
+      }).throws("Cannot assign to read only index 0");
+      expect(() => {
+        //@ts-expect-error Should be an invalid write to read-only object.
+        objects[1] = { doubleCol: 0 };
+      }).throws("Cannot assign to read only index 1");
+      expect(() => {
+        //@ts-expect-error Should be an invalid write to read-only object.
+        objects.length = 0;
+      }).throws("Cannot assign to read only property 'length'");
+    });
+
+    it("should use correct subscript", function (this: RealmContext) {
+      expect(this.realm.objects("TestObject")[0]).equals(undefined);
+
+      this.realm.write(() => {
+        this.realm.create<TestObject>("TestObject", { doubleCol: 1 });
+        this.realm.create<TestObject>("TestObject", { doubleCol: 2 });
+      });
+
+      const results = this.realm.objects<TestObject>("TestObject");
+      expect(results[0].doubleCol).equals(1.0);
+      expect(results[1].doubleCol).equals(2.0);
+      expect(results[2]).equals(undefined);
+      expect(results[-1]).equals(undefined);
+      // TODO: expect(Object.getPrototypeOf(results[0])).equals(TestObject.prototype);
```

Comment:
I am not sure if this is worthwhile but previous test used to do some prototype checks, is this something that should be migrated over?

Reply:
Sure - but it will fail on the `bindgen` branch (since we're injecting a class between the user provided constructor and the user-facing object). I think simply doing an instanceOf check would suffice:

```suggestion
      expect(results[0]).instanceOf(TestObject);
```